docs/guides/python.md

```diff
@@ -0,0 +1,142 @@
+a `test_func.py` file which contains a simple unit test. To run the test locally,
+you'll need to install the required dependencies. You do this as you would
+with any Python project.
+
+```
+❯ pip install -r requirements.txt
+```
+
+Once you have done this, you can run the provided tests with `python3 test_func.py`.
+The default test framework for Python functions is `unittest`. If you prefer another,
+that's no problem. Just install a test framework more to your liking.
+
+## Function reference
+
+Boson Python functions have very few restrictions. You can add any required dependencies
+in `requirements.txt`, and you may include additional local Python files. The only real
+requirements are that your project contain a `func.py` file which contains a `main()` function.
+In this section, we will look in a little more detail at how Boson functions are invoked,
+and what APIs are available to you as a developer.
+
+### Invocation parameters
+
+When using the `func` CLI to create a function project, you may choose to generate a project
+that responds to a `CloudEvent` or simple HTTP. `CloudEvents` in Knative are transported over
+HTTP as a `POST` request, so in many ways, the two types of functions are very much the same.
+They each will listen and respond to incoming HTTP events.
+
+When an inacoming request is received, your function will be invoked with a `Context` object as the first parameter. This object is simply a Python `dict` with two keys, `request` and
+optionally, `cloud_event`, if the incoming request is a `CloudEvent`. The `request` value
+is the Flask request objext. Developers may access any event data from the context objext.
+For example: 
+
+```python
+def main(context: Context):
+  """ 
+  The context parameter contains the Flask request object and any
+  CloudEvent received with the request.
+  """
+  print(f"Method: {context['request'].method}")
+  print(f"Event data {context['cloud_event'].data})
+  # ... business logic here
+```
+
+### Return Values
+Functions may return any value supported by Flask, as the invocation framework
+proxies these values directly to the Flask server. See the Flask 
+[documentation](https://flask.palletsprojects.com/en/1.1.x/quickstart/#about-responses)
+for more information.
+
+#### Example
+```python
+def main(context: Context):
+  attributes = {
+    "type": "com.example.fn",
+    "source": "https://example.com/fn"
+  }
+  data = { "message": "Howdy!" }
+  event = CloudEvent(attributes, data)
+  headers, body = to_binary(event)
+  return body, 200, headers
+```
+
+Note that functions may set both headers and response codes as secondary
+and tertiary response values from function invocation. 
+
+## Dependencies
+Developers are not restricted to the dependencies provided in the template
+`requirements.txt` file. Additional dependencies can be added as they would be
+in any other project by simply adding them to the `requirements.txt` file.
```

Comment:
Can you add an example of what that txt file would look like with a python dependency added just for completeness? Or a command that would add it, do they add it manually, etc?

Reply:
`requirements.txt` could be considered as a standard Python way of describing depenedencies, so every Python dev knows how to manage this file. IMHO I don't think it is neccessary to desribe it for function developers.

Reply:
I agree with @zroubalik this will be fundamental knowledge for any Python dev.

Let's hold this PR until @matejvasek's [PR](https://github.com/boson-project/parliament/pull/9) lands since that will affect the docs for `context`.